ambry-tools/src/main/java/com.github.ambry/tools/admin/ConcurrencyTestTool.java

```diff
@@ -474,6 +508,81 @@ public void onCompletion(Void result, Exception exception) {
   }
 
   /**
+   * Class to assist in collecting and reporting metrics
+   */
+  class MetricsCollector {
+    private String operation;
+    private final long measurementIntervalInSecs;
+    private ArrayList<Long> requestLatencies;
+    private AtomicLong timePassedInNanoSeconds = new AtomicLong(0);
+    private AtomicLong numberOfOperations = new AtomicLong(0);
+    private AtomicLong maxLatencyInNanoSeconds = new AtomicLong(0);
+    private AtomicLong minLatencyInNanoSeconds = new AtomicLong(Long.MAX_VALUE);
+    private AtomicLong totalLatencyInNanoSeconds = new AtomicLong(0);
+
+    MetricsCollector(long measurementIntervalInSecs, String operation) {
+      this.measurementIntervalInSecs = measurementIntervalInSecs;
+      requestLatencies = new ArrayList<Long>();
+      this.operation = operation;
+    }
+
+    /**
+     * Updates the wall clock time spent so far
+     * @param timePassedInNs the spent so far in Nanoseconds
+     */
+    synchronized void updateTimePassedSoFar(long timePassedInNs) {
+      timePassedInNanoSeconds.addAndGet(timePassedInNs);
+      if (timePassedInNanoSeconds.get() >= measurementIntervalInSecs * SystemTime.NsPerSec) {
+        reportMetrics();
+        numberOfOperations.set(0);
+        timePassedInNanoSeconds.set(0);
+        requestLatencies.clear();
+        maxLatencyInNanoSeconds.set(0);
+        minLatencyInNanoSeconds.set(Long.MAX_VALUE);
+        totalLatencyInNanoSeconds.set(0);
+      }
+    }
+
+    /**
+     * Reports the stats for latencies like avg, 95th, 99th, min and max
+     */
+    void reportMetrics() {
+      if (requestLatencies.size() > 5) {
+        Collections.sort(requestLatencies);
```

Comment:
PriorityQueue sorting at insertion vs. List sorting at the end, not sure which one is better, but just throwing out another possibility.